File: Assignment-3/oct03/fortran/pendulum.f90
!  PENDULUM - the forced damped nonlinear pendulum.
!
      module pendulum
      use rkmod
      use,intrinsic:: ieee_arithmetic,only: ieee_rem  ! optional
      implicit none
      real(WP), parameter:: PI=3.1415926535897932_WP
      real(WP), parameter:: TWOPI=2*PI
!
!  Default map parameters
!
      real(WP), save, private:: damp=0.2_WP
      real(WP), save, private:: force=1.66_WP
      contains
!-----------------------------------------------------------------------
      pure subroutine poincare(y, niter)
!  POINCARE - 2-pi Poincare map for the forced damped pendulum.
!
      integer, intent(in):: niter
      real(WP), intent(inout):: y(2)
!
      integer:: k
      integer, parameter:: NSTEPS=256
      real(WP), parameter:: ZERO=0.0

      do k=1,nsteps
         call rk4(forced_damped, 2, y, ZERO, TWOPI, NSTEPS)
         !  Your code here to implement the appropriate modulus
         !  operation(s) on the position variable, y(1)
      enddo
      return
      end subroutine poincare
!-----------------------------------------------------------------------
      pure subroutine forced_damped(y, n, t, dy)
!  Y : 2-vector giving (position, velocity).
!  DY := corresponding time derivatives.
!
      integer, intent(in):: n
      real(WP), intent(in):: y(n), t
      real(WP), intent(out):: dy(n)
!
      if(n.ne.2) then  ! can also be written (n /= 2)
         dy = 0.0
      else
        dy(1) = y(2)
        dy(2) = force*cos(t) - sin(y(1)) - damp*y(2)
      endif
      return
      end subroutine forced_damped
!-----------------------------------------------------------------------
      end module pendulum

File: Assignment-6/pendulum.cc
#include "pendulum.h"
#include "rk4.h"

// This code defines the Poincare' map for the forced damped nonlinear
// pendulum, x'' + damping*x' + sin(x) = force*cos(t).
// Of course, you may use the same sampled implementation as in the C
// code, because C++ is largely upwardly compatible with C.
// This version illustrates the use of a "static class" to structure the
// map and its parameters.

// There is only ever one instance of a static class.
// Static classes do not have constructors; instead, their
// data elements are initialized outside of the class as shown below.

// Any class (not just static classes) can declare a STATIC member function.
// The difference between static member functions and ordinary ones is that
// the static member functions do not have a THIS pointer passed implicitly
// among their arguments.  Consequently, static member functions cannot
// access private data elements unless they are also STATIC.

// The restrictions nevertheless work for this application because
// (1) we integrate exactly one forced, damped pendulum and
// (2) the calling interface to the RK4 procedure does not admit a function
//     containing a pointer to an arbitrary data structure.

// One potential advantage of this arrangement over the C version is that
// the parameters DAMPING and FORCE are in their own namespace.  Outside
// of the member functions, they can be accessed only with the qualified
// names FORCED_DAMPED::DAMPING and FORCED_DAMPED::FORCE and so cannot be
// confused with other variables named DAMPING and FORCE that may be defined
// elsewhere in the program.

inline real_t wrap_around(real_t angle) {
  real_t m = (angle + PI) - TWOPI * floor((angle + PI) / TWOPI);
  return fmod(TWOPI + m, TWOPI) - PI;
}

//  The first-order system that defines the forced, damped pendulum
void forced_damped::pendulum(real_t t, const array_t &y, array_t &dy, size_t num_eqs) {
  for (size_t i = 0; i < num_eqs/2; i++) {
    dy[2 * i] = y[2 * i + 1]; // derivative of position is velocity
    dy[2 * i + 1] = force * std::cos(t) - std::sin(y[2 * i]) - damping * y[2 * i + 1];
  }
  return; // optional
};

// Member function that defines the Poincare' map
void forced_damped::poincare(array_t &y, size_t niter, size_t points_left) {
  const size_t NSTEPS = 256;
  for (size_t j = 0; j < niter; j++) {
    rk4(pendulum, y, 0., TWOPI, NSTEPS, points_left * 2);
    // Your code here to wrap the position, y[0], back into
    // the interval [-PI,PI].
    // https://stackoverflow.com/questions/11980292/how-to-wrap-around-a-range
    for (size_t i = 0; i < points_left; i++)
      y[2 * i] = wrap_around(y[2 * i]);
  }
  return; // optional
};

real_t forced_damped::damping = 0.2;
real_t forced_damped::force = 1.66;
